File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i,n in enumerate(nums):
            if i > 0 and n == nums[i - 1]:
                continue
            j = i + 1
            k = len(nums) - 1
            while j < k:
                s = nums[j] + nums[k] + n
                if s > 0:
                    k -= 1
                elif s < 0:
                    j += 1
                else:
                    res.append([n, nums[j], nums[k]])
                    j += 1
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
        return res

        

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {

    const res = [];
    nums.sort((a, b) => a - b);

    for (let i = 0; i < nums.length - 2; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;
        const target = -nums[i];
        let j = i + 1;
        let k = nums.length - 1;
        while (j < k) {
            const sum = nums[j] + nums[k];
            if (sum === target) {
                res.push([nums[i], nums[j], nums[k]]);
                j++;
                while (j < nums.length && nums[j] === nums[j - 1]) {
                j++;
                }
                k = nums.length - 1;
            } else if (sum > target) k--;
            else j++;
        }
    }


    return res;

};